src/extra/components/adq.py

```diff
@@ -1285,15 +1293,18 @@ def pulse_edge_array(self, labelled=True, squeeze_edges=True,
 
         edge_func = self._validate_edge_method(edge_func, edge_kw)
 
+        # Drop empty trains for efficient access to train IDs.
+        raw_key = self._raw_key.drop_empty_trains()
+
         # Obtain information about pulse layout.
-        pulse_ids, pulse_layout = self._prepare_pulses()
+        pulse_ids, pulse_layout = self._prepare_pulses(raw_key.train_ids)
 
         if max_edges is None:
             max_edges = pulse_layout['length'].max() // 100
 
         # Set-up pasha for parallel processing.
         psh = self._prepare_pasha(parallel)
-        trace_out = np.zeros(self._raw_key.entry_shape, dtype=np.float32)
+        trace_out = np.zeros(raw_key.entry_shape, dtype=np.float32)
```

Comment:
Silly thing, but you assume `float32`. Could you not refer to another `dtype`, so that you keep the same type as the input/